MacOS

rmarinho opened the pull request:
### Description of Change ###

This PR adds support for MacOS backend on Xamarin.Forms.

All the controls are implemented expect OpenGLView, but not all features are present.
We have UITest rudimentary support on CI. 

Opening the PR to prepare for review and merge in the following weeks. As well as get general feedback.
Take in mind most renderers are ported from the iOS counterpart and might have some incongruences that we can now fix in this new implementation , please point them out if see them. 

Xamarin.Forms.Maps.MacOS.Extra is a extra lib that is only needed to support Cycle 8, it has some missing APIS that were only added to Xamarin.Mac in Cycle9.

### Behavioral Changes ###

Add MacOS backend

### PR Checklist ###

- [x] Has tests (if omitted, state reason in description)
- [x] Rebased on top of master at time of PR
- [x] Changes adhere to coding standard
- [ ] Consolidate commits as makes sense

adrianknight89 commented:
Can you please share how you prioritize PRs for merge? There's a lot of open PRs at the moment that fix iOS and Android bugs. I fear that adding another platform is potentially going to open a can of bugs while the two most important platforms need attention.

rmarinho commented:
Open PR's for longer have higher priority.

djcparker commented:
Awesome. Great work, looking forward to it.

StephaneDelcroix commented on Xamarin.Forms.Platform.MacOS/Platform.cs:
```diff
@@ -0,0 +1,262 @@
+
+namespace Xamarin.Forms.Platform.MacOS
+{
+	public class Platform : BindableObject, IPlatform, IDisposable
```
how much of this can be shared? is it worth it ?

rmarinho replied:
I cleaned it a lot i prefer not to share..

StephaneDelcroix commented on Xamarin.Forms.Platform.MacOS/PlatformEffect.cs:
```diff
@@ -0,0 +1,10 @@
+
+namespace Xamarin.Forms.Platform.MacOS
+{
+	public abstract class PlatformEffect : PlatformEffect<NSView, NSView>
```
DRY

StephaneDelcroix commented on Xamarin.Forms.Platform.MacOS/Renderers/EntryRenderer.cs:
```diff
@@ -0,0 +1,167 @@
+			var targetColor = Element.PlaceholderColor;
+
+			// Placeholder default color is 70% gray
+			// https://developer.apple.com/library/prerelease/ios/documentation/UIKit/Reference/UITextField_Class/index.html#//apple_ref/occ/instp/UITextField/placeholder
```
what's the default value for macOS ?

StephaneDelcroix commented on Xamarin.Forms.Platform.MacOS/Renderers/EntryRenderer.cs:
```diff
@@ -0,0 +1,167 @@
+
+namespace Xamarin.Forms.Platform.MacOS
+{
+	public class EntryRenderer : ViewRenderer<Entry, NSTextField>
```
some of this can probably be shared with iOS

rmarinho commented on Xamarin.Forms.Platform.MacOS/Controls/FormsAwesomeBar.cs:
```diff
@@ -0,0 +1,243 @@
+namespace Xamarin.Forms.Platform.MacOS
+{
+	//Ported from Xamarin Studio
+	public class FormsAwesomeBar : NSView
```
sure

rmarinho commented on Xamarin.Forms.Platform.MacOS/NativeToolbarTracker.cs:
```diff
@@ -0,0 +1,221 @@
+			};
+
+			// We can't use the events that Xamarin.Mac adds for delegate methods as they will overwrite
+			// the delegate that Gtk has added
```
was here from Xamarin.Studio

rmarinho commented on Xamarin.Forms.Platform.MacOS/Controls/FormsImageView.cs:
```diff
@@ -0,0 +1,22 @@
+		{
+			_isOpaque = isOpaque;
+		}
+		public override bool IsOpaque
```
Can't do that

rmarinho commented on Xamarin.Forms.Platform.MacOS/Controls/FormsNSView.cs:
```diff
@@ -0,0 +1,59 @@
+			base.DrawRect(dirtyRect);
+		}
+
+		public override bool IsFlipped
```
NSView doesn't have a getter, the property is meant to be overrided

rmarinho commented on Xamarin.Forms.Platform.MacOS/Extensions/NSButtonExtensions.cs:
```diff
@@ -0,0 +1,25 @@
+
+namespace Xamarin.Forms.Platform.MacOS
+{
+	public static class NSButtonExtensions
```
But what if they are used in multiple places ?

rmarinho commented on Xamarin.Forms.Platform.MacOS/ImageSourceHandlers.cs:
```diff
@@ -0,0 +1,67 @@
+using AppKit;
+using Foundation;
+
+namespace Xamarin.Forms.Platform.MacOS
```
not that easy they aren't in the same file , (on iOS are inside the ImageRenderer) i can share this file with iOS, and delete the iOS code ..

AndyDentFree commented on Xamarin.Forms.Platform.MacOS/Controls/FormsScrollView.cs:
```diff
@@ -0,0 +1,58 @@
+			set;
+		}
+
+		public override async void ScrollWheel(NSEvent theEvent)
```
Note that it **is** still called a ScrollWheel event by Apple in [Mouse Events](https://developer.apple.com/library/content/documentation/Cocoa/Conceptual/EventOverview/EventObjectsTypes/EventObjectsTypes.html) especially as there is a virtual scrollwheel on the MagicMouse (and my Windows mice still have them).

However, yes you're right that the OS routes two-finger scrolling gestures on the Trackpack through this event.

As WPF also has a _Wheel_ (Mouse.MouseWheel) I'm in favour of keeping it as named - it keeps the event more aligned with the target platform's framework, which helps people translating code.

AndyDentFree commented:
As a general comment on sharing code, in my experience the more robust process is to get a solid version working **before** trying to mingle code.

It is useful to be _identifying_ sharing opportunities, just don't rush to implement them. When _both_ sides of a framework are actively changing, sharing code becomes a source of conflict, blockage and bugs.

I've worked on a lot of multi-platform frameworks and codebases, and ports between platforms, over the decades. **Please** get a clean v1.0 working then isolate common code.

rmarinho commented:
I have cleaned this branch with only the changes to be reviewed, removed gallery and UITests that could be find if you want to test run the sample on macOS-gallery branch.

samhouts commented on Xamarin.Forms.Platform.MacOS/Renderers/PickerRenderer.cs:
```diff
@@ -0,0 +1,153 @@
+					_disposed = true;
+					if (Element != null)
+					{
+						//((ObservableList<string>)Element.Items).CollectionChanged -= RowsCollectionChanged;
```
Is this a todo?

samhouts commented on Xamarin.Forms.Platform.MacOS/Renderers/ProgressBarRenderer.cs:
```diff
@@ -0,0 +1,67 @@
+
+		protected override void SetBackgroundColor(Color color)
+		{
+			//base.SetBackgroundColor(color);
```
Is this a todo?

samhouts commented on Xamarin.Forms.Platform.MacOS/Renderers/ScrollViewRenderer.cs:
```diff
@@ -0,0 +1,215 @@
+		[Export(nameof(UpdateScrollPosition))]
+		void UpdateScrollPosition()
+		{
+			System.Diagnostics.Debug.WriteLine("need to fix this math");
```
Maybe this should be a todo instead of a debug writeline?

samhouts commented on Xamarin.Forms.Platform.iOS/Renderers/LabelRenderer.cs:
```diff
@@ -144,25 +177,77 @@ void UpdateLineBreakMode()
+			else
+			{
+				Control.StringValue = (string)values[1] ?? "";
+				// default value of color documented to be black in iOS docs
```
What about mac docs? :)

rmarinho replied:
Hum.. didn't found it.

samhouts commented:
I'd recommend putting TODOs on all commented code to make it easier to find.

rmarinho commented on Xamarin.Forms.Platform.MacOS/Platform.cs:
```diff
@@ -0,0 +1,262 @@
+
+		internal NSViewController ViewController => PlatformRenderer;
+
+		internal static void DisposeModelAndChildrenRenderers(Element view)
```
No it's really Model like yourself and children (this is the same on iOS)

rmarinho commented:
@samhouts you were so right and i forgot , but did it now :)
